fedbiomed/common/secagg/_jls.py

```diff
@@ -117,7 +120,14 @@ def divide(xs: List[int], k: int) -> List[int]:
     Returns:
         List of divided integers
     """
-    xs = np.array(xs, dtype=np.uint32)
+
+    # CAVEAT: `dtype` must to allow bigger values than each nodes's weighted value
+    # This implementation allows at most 2**32 nodes (as weighted value uses uint32)
+    max_val = np.iinfo(np.uint64).max
+    if any([v > max_val or v < 0 for v in xs]):
+        raise FedbiomedSecaggCrypterError(f"{ErrorNumbers.FB624.value}: Cannot divide, values outside of bounds")
+
+    xs = np.array(xs, dtype=np.uint64)
```

Comment:
Hi @mvesin,

First of all thank you for fixing this bug. 

I understand that there you want to verify if any value is not larger than max value of `int64`. However, following code will raise an error anyway. 

```
np.array([2**64], dtype=np.int64)
```
Therefore, I propose to use exception rather than looping over values which is not performance efficient.

Reply:
~~Or we can directly use builtin Python types rather than numpy which will avoid getting overflow errors.~~ Nevermind, it is not a good idea because we can get division errors.

Reply:
And I didn't understand why any value of `xs` can't be less than 0.

Reply:
Thank you very much @srcansiz for your review !

> And I didn't understand why any value of xs can't be less than 0.
> OK I see, quantized values are always 0 or positive. So this modification makes this function very specific 
> positive integers or quantized values. I would recommend to leave it more generic.

I see your point ...
The idea was to robustify the computation as (1) `xs` elements are typed as `int` (and should be a positive `int`, but this property is only met if the very complex secure aggregation code has no bugs (2) if `xs` element is negative, this will not be detected when converting to `uint64`. This will result in the same type of silent bug that we are fixing.
Maybe then we should move the test <0 to `SecaggCrypter._apply_average` (which is calling `divide` and is specific to this case) ?

> I understand that there you want to verify if any value is not larger than max value of int64. However, following code will raise an error anyway.

The 2 reasons for adding this test were:
1. overhead is little as most of the computation time is in `aggregate()` not in the `divide()`
2. we can give a more explicit message to the user. I hoped to help identifying and fixing future issues in this piece of code :-) If we remove this test we have something like:

```
File /home/mvesin/GIT/fedbiomed/fedbiomed/fedbiomed/common/secagg/_jls.py:130, in divide(xs, k)
...
--> 130 xs = np.array(xs, dtype=np.uint64)
    131 return (xs / k).tolist()

OverflowError: Python int too large to convert to C long
```

... but if you don't agree we can remove this test.

Reply:
Hi @mvesin, 

Thank you for your answer. I agree that moving these checks to `SecaggCrypter._apply_average` make sense. My point was to keep lower layer functions more generic and deal such invalid types in the upper layers. Because those functions, such as divide, multiply, can be useful for other type of algorithms which may accept negative values. 

For example,  `_apply_average` is a private member of `SecaggCrypter` which calls  `divide` function. There, we can control if is there any negative values or values larger than  max `int64`. This can be done by creating  `np.array, dtype=int64` or looping over the list then it can call `divide` function because it is sure that type and value is correct.

The same argument may be valid for `reverse_quantized`, however, `reverse_quantized` is more specific that it should expect positive values all the time. So it is up to you. 

I am not saying that it is better. It is just to take your opinion about it.
 
> overhead is little as most of the computation time is in aggregate() not in the divide()

Regarding the performance, you are right that it does not take much time compare to aggregation or encryption. My point is just looping over a list and converting list to numpy array takes more or less same amount of time. In this implementation both of them are existing. However, 1M parameters won't make any difference but when you go up to 40M (e.g UNET). It takes around 1.5 seconds, plus 1.5 for converting it to `np.array` 3 seconds in total, same operation for `reverse_quantized` 6 seconds in total. However, I know that there many such loops in secure aggregation implementation so changing this one doesn't make much difference :)

Reply:
> My point was to keep lower layer functions more generic and deal such invalid types in the upper layers. 

OK, so I will move the test for `multiply()` (to be added as suggested by @sharkovsky ) and `divide()` to `SecaggCrypter`.

For `reverse_quantize()` I would keep it as it is, for the reason you mention (specific function for this processing).

> Regarding the performance

I would like to keep these checks, for robustness sake.

We can gain performance by removing the conversion to numpy: as the treatment is simple (divide/multiply) this is more costy to convert twice from `int => numpy => int` than do raw python computation

```
a = list(range(100*10**6))

before = datetime.now() ; b = divide(a,7) ; print(datetime.now() - before)
0:00:06.745853

def divide(xs: List[int], k: int) -> List[int]:
    return ([e / k for e in xs])

before = datetime.now() ; b = divide(a,7) ; print(datetime.now() - before)
0:00:04.025618
```

Reply:
If we plan to re-use the `quantize` / `reverse_quantize` and `multiply` / `divide` it might be nice to re-implement fully in `numpy` (so that we only convert / unconvert once in the process, and then do full numpy computation). But this is beyond the scope of this fix, maybe for (next ?) secagg sprint.